lib/spack/spack/paths.py

```diff
@@ -9,135 +9,318 @@
 dependencies.
 """
 import os
+import pathlib
+from enum import Enum
 from pathlib import PurePath
 
 import llnl.util.filesystem
 
-#: This file lives in $prefix/lib/spack/spack/__file__
-prefix = str(PurePath(llnl.util.filesystem.ancestor(__file__, 4)))
+import spack.util.hash as hash
+
+
+class XDG_vars(Enum):
+    config_home = "XDG_CONFIG_HOME"
+    state_home = "XDG_STATE_HOME"
+    data_home = "XDG_DATA_HOME"
+    cache_home = "XDG_CACHE_HOME"
+
+
+class Location_vars(Enum):
+    user_cache_path = "USER_CACHE_PATH"
+    spack_data_home = "SPACK_DATA_HOME"
+
+
+# This is for tests that want to clean the environment of XDG_ variables that
+# affect spack behavior
+def _unset_xdg_vars(env):
+    saved = {}
+    for xdg_var in XDG_vars:
+        if xdg_var.value in env:
+            saved[xdg_var.value] = env.pop(xdg_var.value)
+    return saved
+
+
+def _define_xdg_or_backup(xdg_var, backup):
+    if xdg_var in os.environ:
+        spack_xdg_defined = os.path.join(os.environ[xdg_var], "spack")
+    else:
+        spack_xdg_defined = os.path.join(backup, "spack")
+    return os.path.expanduser(spack_xdg_defined)
+
+
+def dir_is_occupied(x, except_for=None):
+    x = pathlib.Path(x)
+    except_for = except_for or set()
+    return x.is_dir() and bool(set(x.iterdir()) - except_for)
+
+
+class SpackPaths:
+    def __init__(self, _prefix=None):
+        #: This file lives in $prefix/lib/spack/spack/__file__
+        self.prefix = _prefix or str(PurePath(llnl.util.filesystem.ancestor(__file__, 4)))
+
+        #: synonym for prefix
+        self.spack_root = self.prefix
+
+        #: bin directory in the spack prefix
+        self.bin_path = os.path.join(self.prefix, "bin")
+
+        #: The spack script itself
+        self.spack_script = os.path.join(self.bin_path, "spack")
+
+        #: The sbang script in the spack installation
+        self.sbang_script = os.path.join(self.bin_path, "sbang")
```

Comment:
I believe this location has been deprecated for several years now, we should consider removing it for 1.0

Reply:
Sounds like this is a later thing: I'm going to mark this as resolved.